Bring the format tutorial into the Format documentation

johnwhitington opened the pull request:
This PR brings the Format tutorial http://caml.inria.fr/resources/doc/guides/format.en.html into the Format module's documentation. This tutorial is already linked from the Format page.

Rationale:

- The Format module has been a pain point for many people. Learning to use it just from the reference manual's description is difficult for most people.
- The tutorial has been around for a long time, and is already blessed by being linked to from the page.
- It adds length to an already-long documentation page, but its usefulness outweighs this.
- Reference manuals can have little tutorials in them, where appropriate. Here is an appropriate place.

I have preserved the formatting of the HTML original as much as possible. The only change I have made is to remove references to Caml Light.

This is not ready to be merged -- I would like to have another careful look at it, due to the complication of removing all the HTML tags and adding the formatting back in in OCamldoc -- but I would appreciate a decision on whether this is a good idea or not.

cc @pierreweis

johnwhitington committed:
Add the tutorial from caml.inria.fr into Format's documentation

johnwhitington committed:
check-typo fixes

johnwhitington committed:
Spacing

Octachron commented:
I have some reservations about this tutorial. There are some points that definitively need some editing. I also think that @Drup will want to comment too.

Octachron commented on stdlib/format.mli:
```diff
@@ -107,6 +103,446 @@
+      pretty-printing engine that proposes to break the line here, if it is
+      necessary to properly print the rest of the material.
+      Otherwise, the pretty-printing engine never break lines (except
+      "in case of emergency" to avoid very bad output).
```
Opening a box after the maximum indentation limit also induces breaks the line. It is a far too common situation to speak about emergency.

gasche commented:
Could you upload a HTML version of the resulting page somewhere, to make it easier for people to have a quick look?

Octachron commented on stdlib/format.mli:
```diff
@@ -107,6 +103,446 @@
+
+Let me give an example. Suppose we can write 10 chars before
+the right margin (that indicates no more room). We represent any
+char as a [-] sign; characters [\[] and [\]]
```
Rather than introducing a purely symbolic DSL, I would recommend to use format strings.
For instance, using `--@ --@ --` instead of `--b--b--` would make it easier to follow the tutorial in a toplevel.

gasche commented:
There are two questions here: (1) whether the tutorial is inside/outside the reference manual and (2) how to improve its contents. I think that the two are somewhat orthogonal (as John mentioned, the tutorial is already referenced explicitly in the manual; we want to improve its content in any case). But then, if this PR is the right excuse for @Octachron and @Drup to provide valuable feedback on the tutorial, and @johnwhitington is willing to turn that feedback into actual changes, well this is excellent news -- even if it means that decisions/integration will be slower for a PR dedicated only to the placement question.

Octachron commented on stdlib/format.mli:
```diff
@@ -107,6 +103,446 @@
+(---
+ (----
+  (---)))
+]}
```
This example and the one below are confusing since there is no break hints before the inner `(`s, the hov boxes are supposed to be open without any indentations. In fact, the only way to get the expected output is to add some indentation and twiddle with the maximum indentation limit (which is not explained at all in this tutorial!):
```ocaml
Format.set_geometry ~max_indent:3 ~margin:10;
Format.printf 
"@[<hov 1>(---@[<hov 1>(----@[<hov 1>(---@,)@]@,)@,@]@,)@]@.";;
```

Octachron replied:
Note that this remark can be fixed by updating the introduction to mention the box indentation and adding break hints before opening the boxes rather than relying on the maximum indentation:

> For instance, when printing
"\[(---b\[(----b\[(---b)\]b)\]b)\]", where "b" is a break hint without extra
indentation ([print_cut ()]). If "\[" means opening of a packing "hov" box
with an indentation of 1,
({!open_hovbox}), "\[(---b\[(----b\[(---b)\]b)\]b)\]" is printed as follows:

or using format string

> For instance, when printing
"@[<hov 1>(---@,@[<hov 1>(----@,@[<hov 1>(---@,)@]@,)@]@,)@]"

Octachron commented on stdlib/format.mli:
```diff
@@ -107,6 +103,446 @@
+
++ Boxes must be opened and closed consistently ([open_*] and {!close_box} must
+be nested like parentheses).
++ Never hesitate to open a box.
```
This advice does not work in my experience. It may lead people to write the following `list` function:
```ocaml
let rec list elt ppf = function
| [] -> ()
| a :: q -> Format.fprintf ppf "%a,@ @[%a@]" elt a (list elt) q
```

johnwhitington commented:
Here is the HTML output:

https://coherentpdf.com/tmp/Format.html

Drup commented:
After a little bit of thinking, I have two main issues with this tutorial, one easily solvable, the other needs some discussion.

First, I'm not fond in putting fully-fledged tutorials inside API documentation. It takes a lot of space, it makes it less easy to navigate the documentation for both newcomers and experienced programmers. We should turn the tutorial into an independent `.mld` page that is part of the stdlib documentation, but in a different page, and link its relevant sections everywhere appropriate.

My main issue is that this tutorial is fundamentally flawed, as it encourages using the main `Format` API, which is really really bad. [This tutorial](https://blag.cedeela.fr/format-all-the-data-structures/) is already a better introduction by the single virtue that it uses a decent API.  
I was happy to turn a blind eye on the maybe-official-but-not-sure tutorial, and only recommend tutorials using a better API. but if we want to integrate and improve it, then maybe its time to improve the `Format` API, or accept that `Format` is barely usable without external libraries and rewrite the tutorial in consequence.  
Regardless of the choice we make, 1) we should make it now, to pave the way forward 2) I'm happy merging the tutorial in an unsatisfying state, and work towards a better API (and tutorial!) later, incrementally.

One argument I can give towards finally fleshing out `Format` is that `Fmt` and `CCFormat`'s API have a 80% surface overlap and have been almost completely stable for the last 5 years. We can almost take their common part directly. The main reason I haven't tried to do it already is lack of will/time. :)

gasche commented:
(With apologies for the not-so-helpful feedback): After looking at the rendered output, I find the increase in length of the module documentation a bit unpleasant: I use it mostly as a reference (to check whether various functions exist at the name I expect, etc., maybe go check the prototype of `pp_print_list`), and that's a substantial amount of extra text that I have to scroll past to get what I was looking for.

johnwhitington commented:
Thanks for all the useful comments. Once they die down, perhaps at the weekend, I will summarise them and suggest some ways forward -- and then we can pick one.

xavierleroy commented:
This is a tough case!  On the one hand, I agree that this PR proposes useful additions to the documentation of the Format module.  In particular, the graphical illustrations of the various kinds of boxes are very helpful.  On the other hand, I agree that the tutorial at http://caml.inria.fr/resources/doc/guides/format.en.html is not that good and could use a rewrite to begin with.

johnwhitington commented:
There are three possible conclusions about the tutorial:

1. It's good
2. It's mediocre but doesn't harm anyone
3. It's actively harmful

It appears we are currently at (3). Can we get to (2) with some simple cuts/additions/changes? If so, I would be happy to coordinate that, but the actual fixes will have to be devised by others.

I agree with @drup and @gasche that this is best moved to another page, merely linked from the Format page.

So, unless anyone is in favour of simply moving it to another page and merging immediately (on the basis that this is not much worse than linking to it) this PR stalls until concrete fixes can be suggested.

Drup commented:
I think the section of the tutorial which explains the boxing model is already quite decent, especially with @Octachron's proposed improvements, notably using the formatting string's syntax directly.

Octachron commented:
I tend to agree that the core of the tutorial is interesting, and after removing the few misleading parts, it is probably a good start. Moreover, it would be easier to update this tutorial as a part of the manual.

johnwhitington committed:
Move format tutorial onto its own page

johnwhitington commented:
I have moved the format tutorial onto its own page, by creating api_docgen/Format_tutorial.mld, extending Makefile.docfiles, and linking to it from stdlib/format.mli.

I am not quite happy with the formatting of the top of the page, or the link to it, but let's fix the content first.

johnwhitington committed:
Merge branch 'trunk' of https://github.com/ocaml/ocaml into format-tutorial

johnwhitington committed:
Update formatting, remove link for now

johnwhitington commented:
I have fixed the formatting on the new Format_tutorial.mld (minus the title, which appears to be automatically generated as the ugly "Format_tutorial", and which I do not know how to fix -- anyone?).

I have removed the link to the format tutorial from the Format page, for now.

I suggest we merge this PR, which adds the tutorial .mld source, and the Makefile changes for it. Then,  those who know Format well can fix up the tutorial in one or more separate PRs (I am willing to coordinate, of course, but know nothing about Format). Then we can reinstate the link once we are sure the tutorial is not actively harmful.

johnwhitington commented:
Rendered version here: https://coherentpdf.com/tmp/Format_tutorial.html

johnwhitington committed:
Tidying up

Octachron reviewed (approved):
I agree that we can advance by steps, and there should be enough time to fix the most egregious quirks of the tutorial in time for the 4.13 release.
